File: examples/src/main/java/de/maibornwolff/domainchoreograph/examples/statistik/domaintypes/GesamtAuswertung.java
package de.maibornwolff.domainchoreograph.examples.statistik.domaintypes;

import de.maibornwolff.domainchoreograph.core.api.DomainDefinition;
import de.maibornwolff.domainchoreograph.core.api.DomainFunction;
import de.maibornwolff.domainchoreograph.examples.statistik.choreographies.StatistikCalculator;

import java.util.List;
import java.util.stream.Collectors;

@DomainDefinition
public class GesamtAuswertung {

    private List<Statistik> statistikList;

    public GesamtAuswertung(List<Statistik> statistikList) {
        this.statistikList = statistikList;
    }

    @DomainFunction
    public static GesamtAuswertung resolve(StatistikCalculator statistikCalculator, StichprobenPool datenpool) {
        List<Statistik> statistikList = datenpool.getStichprobenList().stream()
            .map(statistikCalculator::calculate)
            .collect(Collectors.toList());

        return new GesamtAuswertung(statistikList);
    }

    public List<Statistik> getStatistikList() {
        return statistikList;
    }

    @Override
    public String toString() {
        return "Anzahl Gesamtauswertungen: " + statistikList.size();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        GesamtAuswertung that = (GesamtAuswertung) o;

        return statistikList.equals(that.statistikList);
    }

    @Override
    public int hashCode() {
        return statistikList.hashCode();
    }
}

File: core/src/test/java/de/maibornwolff/domainchoreograph/example/domaindefinitions/GesamtAuswertung.kt
package de.maibornwolff.domainchoreograph.example.domaindefinitions

import de.maibornwolff.domainchoreograph.core.api.DomainDefinition
import de.maibornwolff.domainchoreograph.core.api.DomainFunction
import de.maibornwolff.domainchoreograph.example.choreographies.StatistikChoreography
import de.maibornwolff.domainchoreograph.example.domaindefinitions.statistik.Statistik
import de.maibornwolff.domainchoreograph.example.domaindefinitions.statistik.Stichprobe

@DomainDefinition
data class GesamtAuswertung(val statistikList: List<Statistik>) {

    override fun toString(): String {
        return "Anzahl Gesamtauswertungen: " + statistikList.size
    }

    companion object {

        @DomainFunction
        fun resolve(statistikChoreography: StatistikChoreography, datenpool: Stichproben): GesamtAuswertung {

            val calculateStatistik = calculateStatistikFunction(statistikChoreography)

            val statistikList = datenpool.stichprobenList
                .map { stichprobe -> calculateStatistik(stichprobe) }
                .filterNotNull()

            return GesamtAuswertung(statistikList)
        }

        private fun calculateStatistikFunction(statistikChoreography: StatistikChoreography): (Stichprobe) -> Statistik? {
            return { stichprobe: Stichprobe ->
                try {
                    statistikChoreography.calculate(stichprobe)
                } catch (exception: IllegalArgumentException) {
                    null
                }
            }
        }
    }
}
